src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,89 @@
+        Field[] declaredFields = clazz.getDeclaredFields();
+
+        for (Field declaredField : declaredFields) {
+            if (declaredField.isAnnotationPresent(Inject.class)) {
+                Object fieldInstance = getInstance(declaredField.getType());
+
+                clazzImplementationInstance = createNewInstance(clazz);
+
+                declaredField.setAccessible(true);
+                try {
+                    declaredField.set(clazzImplementationInstance, fieldInstance);
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException("Can't initialize field value. "
+                            + "class: " + clazz.getName() + ".Field: "
+                            + declaredField.getName(), e);
+                }
+            }
+        }
+        if (clazzImplementationInstance == null) {
+            clazzImplementationInstance = createNewInstance(clazz);
+        }
+        return clazzImplementationInstance;
+    }
+
+    private Object createNewInstance(Class<?> clazz) {
+        if (instances.containsKey(clazz)) {
+            return instances.get(clazz);
+        }
+
+        try {
+            Constructor<?> constructor = clazz.getConstructor();
+            Object instance = constructor.newInstance();
+            instances.put(clazz, instance);
+            return instance;
+        } catch (InstantiationException | InvocationTargetException | IllegalAccessException
+                 | NoSuchMethodException e) {
+            throw new RuntimeException("can't create new instance of: " + clazz.getName(), e);
+        }
+    }
+
+    private Class<?> findImplementation(Class<?> interfaceClazz) {
+        Map<Class<?>, Class<?>> interfaceImplementations = new HashMap<>();
+        String packageName = "mate.academy.service.impl";
+        String packagePath = packageName.replace('.', '/');
+
+        File packageDirectory = new File(Thread.currentThread()
+                .getContextClassLoader().getResource(packagePath).getFile());
+
+        for (File file : packageDirectory.listFiles()) {
+            if (file.isFile() && file.getName().endsWith(".class")) {
+                String className = packageName + "."
+                        + file.getName().replace(".class", "");
+                try {
+                    Class<?> clazz = Class.forName(className);
+                    if (!clazz.isInterface() && !Modifier.isAbstract(clazz.getModifiers())) {
+                        if (interfaceClazz.isAssignableFrom(clazz)) {
+                            interfaceImplementations.put(interfaceClazz, clazz);
+                        }
+                    }
+                } catch (ClassNotFoundException e) {
+                    throw new RuntimeException("there is no Implementations to this interface: "
+                            + interfaceClazz, e);
+                }
+            }
+        }
+        return interfaceImplementations.get(interfaceClazz);
```

Comment:
Looks interesting. I don't know if we need to do it this way (because there is a much simpler implementation), but ok

Reply:
Looks very cool, good job!
An improvement idea that I have in mind for this is we can try to update this method and call it in the static initialization block, and try to scan and find implementations for the whole project there, to have that `interfaceImplementations` map filled before we even call this `findImplementation()`.
Also, there is no check that impl class is marked with component annotation, we should definitely add it somewhere here.
But of course, all of that is extra work and not necessary at all in this task, just for your free time :)